app/forms/ctc/mailing_address_form.rb

```diff
@@ -1,15 +1,35 @@
 module Ctc
   class MailingAddressForm < QuestionsForm
     set_attributes_for :intake, :street_address, :street_address2, :state, :city, :zip_code
+    set_attributes_for :misc, :address_not_found
 
     validates_presence_of :street_address
     validates_presence_of :city
     validates_presence_of :state
     validates :zip_code, zip_code: true
-
+    validate :usps_valid_address
 
     def save
       @intake.update(attributes_for(:intake))
     end
+
+    private
+
+    def usps_valid_address
+      return unless errors.blank?
+
+      @intake.assign_attributes(attributes_for(:intake))
+      address_service = StandardizeAddressService.new(@intake)
+      if !address_service.valid?
+        case address_service.error_code
+        when "USPS-2147219400"
+          errors.add(:city, I18n.t("forms.errors.mailing_address.city"))
+        when "USPS-2147219402"
+          errors.add(:state, I18n.t("forms.errors.mailing_address.state"))
+        when "USPS-2147219403", "USPS-2147219401", "USPS-2147219396"
+          errors.add(:address_not_found, "Error: Address Not Found")
```

Comment:
thinking these error codes / attached fields+errors should be in their own data structure of some sort but not sure how best to do it. maybe something like:
``` 
[
  {
    code: ""USPS-2147219400"",
    attr: :city,
    error: I18n.t("forms.errors.mailing_address.city")
  }
]
```
or
```
{
  "USPS-2147219400": {
    attr: :city,
    error: I18n.t("forms.errors.mailing_address.city")
  }
}
```
not sure whether the I18n function will work in a constant...

Reply:
I wonder if you could move all of this logic into the translations themselves, especially given that they're each tied to specific error codes that could be used as keys:

````
forms:
  errors:
     mailing_address:
        USPS-2147219402: "Error: Invalid State"
        USPS-2147219402: "Error: Invalid City"
        USPS-2147219403: &generic_address_error "Error: Invalid Address"
        USPS-2147219401: *generic_address_error
        ... etc
```

This falls down if there is a new or different error code that we get back from the IRS API, but I'm fairly certain you could create a fallback pretty easily.

Reply:
I wonder if you could move all of this logic into the translations themselves, especially given that they're each tied to specific error codes that could be used as keys:

````
forms:
  errors:
     mailing_address:
        USPS-2147219402: "Error: Invalid State"
        USPS-2147219402: "Error: Invalid City"
        USPS-2147219403: &generic_address_error "Error: Invalid Address"
        USPS-2147219401: *generic_address_error
        ... etc
```

This falls down if there is a new or different error code that we get back from the IRS API, but I'm fairly certain you could create a fallback pretty easily.